kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/internal/expressions/DefaultExpressionEvaluatorSuite.scala

```diff
@@ -773,6 +773,186 @@ class DefaultExpressionEvaluatorSuite extends AnyFunSuite with ExpressionSuiteBa
+        new DefaultExpressionEvaluator(
+          schema, substringExpression, StringType.STRING).eval(input)
+      val expOutputVector = stringVector(expOutputSeq);
+      checkStringVectors(actOutputVector, expOutputVector)
+    }
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 0),
+        Seq[String](null, "one", "two", "three", "four", null, null, "seven", "eight"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 1),
+        Seq[String](null, "one", "two", "three", "four", null, null, "seven", "eight"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2),
+        Seq[String](null, "ne", "wo", "hree", "our", null, null, "even", "ight"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -1),
+        Seq[String](null, "e", "o", "e", "r", null, null, "n", "t"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -1000),
+        Seq[String](null, "one", "two", "three", "four", null, null, "seven", "eight"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 0, Option(4)),
+        Seq[String](null, "one", "two", "thre", "four", null, null, "seve", "eigh"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2, Option(0)),
+        Seq[String](null, "", "", "", "", null, null, "", ""))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2, Option(1)),
+        Seq[String](null, "n", "w", "h", "o", null, null, "e", "i"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2, Option(10000)),
+        Seq[String](null, "ne", "wo", "hree", "our", null, null, "even", "ight"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 1000),
+        Seq[String](null, "", "", "", "", null, null, "", ""))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), 2, Option(-10)),
+        Seq[String](null, "", "", "", "", null, null, "", ""))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -2, Option(1)),
+        Seq[String](null, "n", "w", "e", "u", null, null, "e", "h"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -2, Option(2)),
+        Seq[String](null, "ne", "wo", "ee", "ur", null, null, "en", "ht"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -4, Option(3)),
+        Seq[String](null, "on", "tw", "hre", "fou", null, null, "eve", "igh"))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -100, Option(95)),
+        Seq[String](null, "", "", "", "", null, null, "", ""))
+
+      checkSubString(
+        input,
+        substring(new Column(col_name), -100, Option(98)),
+        Seq[String](null, "o", "t", "thr", "fo", null, null, "sev", "eig"))
+
+      checkSubString(
```

Comment:
case where the start index is beyond the input string length?

Reply:
checkSubString(
        input,
        substring(new Column(col_name), 1000),
        Seq[String](null, "", "", "", "", null, null, "", "")) above handle this case, also added two test cases for testing Interger.MAX and Integer.MIN